File: src/lib/choice-label-format.ts
import { normalizeAnswerForGrade } from "./answer-normalize";

const CIRCLE_NUMBERS = ["①", "②", "③", "④", "⑤", "⑥", "⑦", "⑧", "⑨", "⑩"];

function escapeRegex(text: string): string {
  return text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/** 선지 label 앞의 "1. ", "1) ", "(1) ", "①" 등 중복 번호 제거 */
export function stripLeadingChoiceMarker(
  label: string,
  choiceId?: string,
): string {
  let text = label.trim();
  if (!text) return text;

  const id = choiceId?.trim();
  if (id) {
    const num = Number.parseInt(id, 10);
    if (num >= 1 && num <= CIRCLE_NUMBERS.length) {
      const circle = CIRCLE_NUMBERS[num - 1]!;
      if (text.startsWith(circle)) {
        text = text.slice(circle.length).trim();
      }
    }

    const escaped = escapeRegex(id);
    const idPatterns = [
      new RegExp(`^\\(?${escaped}\\)?\\.\\s+`),
      new RegExp(`^\\(?${escaped}\\)?\\)\\s+`),
      new RegExp(`^${escaped}:\\s+`),
      new RegExp(`^${escaped}\\s+(?!\\d)`),
    ];
    for (const pattern of idPatterns) {
      if (pattern.test(text)) {
        text = text.replace(pattern, "").trim();
        break;
      }
    }
  }

  text = text.replace(/^[①②③④⑤⑥⑦⑧⑨⑩]\s*/, "");
  text = text.replace(/^\(?[1-9]\)?[.):：、]\s+/, "");

  return text;
}

export function formatChoiceDisplayLabel(id: string, label: string): string {
  return `${id}. ${stripLeadingChoiceMarker(label, id)}`;
}

export function sanitizeProblemChoices<T extends { id: string; label: string }>(
  choices: T[] | undefined,
): T[] | undefined {
  if (!choices?.length) return choices;
  return choices.map((choice) => ({
    ...choice,
    label: stripLeadingChoiceMarker(choice.label, choice.id),
  }));
}

/** 객관식 선지·정답 문자열에서 중복 번호 제거 */
export function normalizeMultipleChoiceProblem<
  T extends {
    type: string;
    correctAnswer: string;
    choices?: { id: string; label: string }[];
  },
>(problem: T): { problem: T; changed: boolean } {
  if (problem.type !== "multiple_choice" || !problem.choices?.length) {
    return { problem, changed: false };
  }

  let changed = false;
  const choices = problem.choices.map((choice) => {
    const label = stripLeadingChoiceMarker(choice.label, choice.id);
    if (label !== choice.label) changed = true;
    return { ...choice, label };
  });

  let correctAnswer = problem.correctAnswer;
  const trimmed = correctAnswer.trim();
  const byId = choices.find((choice) => choice.id === trimmed);
  if (!byId) {
    for (const choice of choices) {
      const strippedCorrect = stripLeadingChoiceMarker(trimmed, choice.id);
      if (
        normalizeAnswerForGrade(strippedCorrect) ===
        normalizeAnswerForGrade(choice.label)
      ) {
        if (correctAnswer !== choice.label) {
          correctAnswer = choice.label;
          changed = true;
        }
        break;
      }
    }
  }

  return {
    problem: { ...problem, choices, correctAnswer },
    changed,
  };
}

File: flutter_app/lib/utils/choice_label_format.dart
/// 선지 label 앞의 "1. ", "1) ", "(1) ", "①" 등 중복 번호 제거
String stripLeadingChoiceMarker(String label, [String? choiceId]) {
  var text = label.trim();
  if (text.isEmpty) return text;

  const circles = ['①', '②', '③', '④', '⑤', '⑥', '⑦', '⑧', '⑨', '⑩'];
  final id = choiceId?.trim();
  if (id != null && id.isNotEmpty) {
    final num = int.tryParse(id);
    if (num != null && num >= 1 && num <= circles.length) {
      final circle = circles[num - 1];
      if (text.startsWith(circle)) {
        text = text.substring(circle.length).trim();
      }
    }

    final patterns = [
      RegExp('^\\(?${RegExp.escape(id)}\\)?\\.\\s+'),
      RegExp('^\\(?${RegExp.escape(id)}\\)?\\)\\s+'),
      RegExp('^${RegExp.escape(id)}:\\s+'),
      RegExp('^${RegExp.escape(id)}\\s+(?!\\d)'),
    ];
    for (final pattern in patterns) {
      if (pattern.hasMatch(text)) {
        text = text.replaceFirst(pattern, '').trim();
        break;
      }
    }
  }

  text = text.replaceFirst(RegExp(r'^[①②③④⑤⑥⑦⑧⑨⑩]\s*'), '');
  text = text.replaceFirst(RegExp(r'^\(?[1-9]\)?[.):：、]\s+'), '');
  return text;
}

/// UI 표시용 — label만 쓰고 id 인덱스(1. 2. …)는 붙이지 않음
String formatChoiceDisplayLabel(String id, String label) {
  return stripLeadingChoiceMarker(label, id);
}
